src/OpticalFlowEstimator.cpp

```diff
@@ -0,0 +1,609 @@
+        }
+    } else {
+        expected_input_size_ = curr_image.size();
+        last_scaled_image_.upload(curr_image);
+        ROS_ASSERT(onSettingsChanged());
+        have_valid_last_image_ = true;
+    }
+
+    last_message_time_ = message->header.stamp;
+    last_orientation_ = current_orientation_;
+    last_camera_to_level_quad_tf_ = current_camera_to_level_quad_tf_;
+}
+
+bool OpticalFlowEstimator::waitUntilReady(
+        const ros::Duration& startup_timeout)
+{
+    return updateFilteredPosition(ros::Time::now(), startup_timeout);
+}
+
+geometry_msgs::TwistWithCovarianceStamped
+        OpticalFlowEstimator::estimateVelocity(
+                const cv::Point2f& average_vec,
+                const tf2::Quaternion& curr_orientation,
+                const tf2::Quaternion& last_orientation,
+                const ros::Time& time) const
+{
+    // Get the pitch and roll of the camera in euler angles
+    // NOTE: CAMERA FRAME CONVENTIONS ARE DIFFERENT, SEE REP103
+    // http://www.ros.org/reps/rep-0103.html
+    double yaw, pitch, roll;
+    getYPR(curr_orientation, yaw, pitch, roll);
+
+    double last_yaw, last_pitch, last_roll;
+    getYPR(last_orientation, last_yaw, last_pitch, last_roll);
+
+    // Calculate time between last and current frame
+    double dt = (time - last_message_time_).toSec();
+
+    // Publish the orientation we're using for debugging purposes
+    iarc7_msgs::OrientationAnglesStamped ori_msg;
+    ori_msg.header.stamp = time;
+    ori_msg.data.pitch = pitch;
+    ori_msg.data.roll = roll;
+    ori_msg.data.yaw = yaw;
+    orientation_pub_.publish(ori_msg);
+
+    // Distance from the camera to the ground plane, along the camera's +z axis
+    double distance_to_plane = current_altitude_
+                             * std::sqrt(1.0
+                                       + std::pow(std::tan(pitch), 2.0)
+                                       + std::pow(std::tan(roll), 2.0));
+
+    // Converts measurements in pixels to measurements in meters in the plane
+    // parallel to the camera's xy plane and going through the point Pg, where
+    // Pg is intersection between the camera's +z axis and the ground plane
+    double current_meters_per_px = distance_to_plane
+                         / getFocalLength(target_size_,
+                                          flow_estimator_settings_.fov);
+
+    // Calculate the average velocity in the level camera frame (i.e. camera
+    // frame if our pitch and roll were zero)
+    // Note that cos(roll) is negative, because the z axis is down
+    cv::Point2f velocity_uncorrected;
+    velocity_uncorrected.x = (-average_vec.x * current_meters_per_px)
+                           / std::cos(pitch)
+                           / dt;
+    velocity_uncorrected.y = (-average_vec.y * current_meters_per_px)
+                           / -std::cos(roll)
+                           / dt;
+
+    double dp;
+    double dr;
+
+    if (last_pitch > CV_PI/2 && pitch < -CV_PI/2) {
```

Comment:
We should have a comment about why this is happening.

Reply:
Done